corva/api.py

```diff
@@ -0,0 +1,129 @@
+    """
+
+    TIMEOUT_LIMITS = (3, 30)  # seconds
+
+    def __init__(
+        self,
+        *,
+        api_url: str,
+        data_api_url: str,
+        api_key: str,
+        app_name: str,
+        timeout: Optional[int] = None,
+    ):
+        self.api_url = api_url
+        self.data_api_url = data_api_url
+        self.api_key = api_key
+        self.app_name = app_name
+        self.timeout = timeout or self.TIMEOUT_LIMITS[1]
+
+    @property
+    def default_headers(self):
+        return {
+            'Authorization': f'API {self.api_key}',
+            'X-Corva-App': self.app_name,
+        }
+
+    def get(self, path: str, **kwargs):
+        return self._request('GET', path, **kwargs)
+
+    def post(self, path: str, **kwargs):
+        return self._request('POST', path, **kwargs)
+
+    def patch(self, path: str, **kwargs):
+        return self._request('PATCH', path, **kwargs)
+
+    def put(self, path: str, **kwargs):
+        return self._request('PUT', path, **kwargs)
+
+    def delete(self, path: str, **kwargs):
+        return self._request('DELETE', path, **kwargs)
+
+    def _get_url(self, path: str):
+        """Builds complete url.
+
+        Args:
+          path: either complete or only path part of the HTTP URL.
+
+        Returns:
+          1 path param, if path is a complete url.
+          2 data api url, if path contains data api url pattern.
+          3 corva api url, if above points are False.
+        """
+
+        if path.startswith('http'):
+            return path
+
+        path = path.lstrip(
+            '/'
+        )  # delete leading forward slash for posixpath.join to work correctly
+
+        # search text like api/v1 or api/v10 in path
+        if bool(re.search(r'api/v\d+', path)):
+            return posixpath.join(self.data_api_url, path)
+
+        return posixpath.join(self.api_url, path)
+
+    def _request(
+        self,
+        method: str,
+        path: str,
+        *,
+        data: Optional[dict] = None,
+        params: Optional[dict] = None,
+        headers: Optional[dict] = None,
+        timeout: Optional[int] = None,
+    ) -> requests.Response:
+        """Executes the request.
+
+        Args:
+          method: HTTP method.
+          path: url to call.
+          data: request body, that will be casted to json.
+          params: url query string params.
+          headers: additional headers to include in request.
+          timeout: custom request timeout in seconds.
+
+        Returns:
+          requests.Response instance.
+        """
+
+        timeout = timeout or self.timeout
+
+        url = self._get_url(path)
+
+        headers = {
+            **self.default_headers,
+            **(headers or {}),
+        }
+
+        response = requests.request(
+            method=method,
+            url=url,
+            params=params,
+            json=data,
+            headers=headers,
+            timeout=self._validate_timeout(timeout),
```

Comment:
Validation should happen prior to usage. Can this be moved up? Also, i don't think it needs to return the timeout, just run the validate routine and raise an error.